consensus/state_processing/src/block_replayer.rs

```diff
@@ -0,0 +1,292 @@
+    ///
+    /// This can be used to inspect the state as blocks are applied.
+    pub fn post_block_hook(mut self, hook: PostBlockHook<'a, E, Error>) -> Self {
+        self.post_block_hook = Some(hook);
+        self
+    }
+
+    /// Run a function immediately before slot processing advances the state to the next slot.
+    pub fn pre_slot_hook(mut self, hook: PreSlotHook<'a, E, Error>) -> Self {
+        self.pre_slot_hook = Some(hook);
+        self
+    }
+
+    /// Run a function immediately after slot processing has advanced the state to the next slot.
+    ///
+    /// The hook receives the state and a bool indicating if this state corresponds to a skipped
+    /// slot (i.e. it will not have a block applied).
+    pub fn post_slot_hook(mut self, hook: PostSlotHook<'a, E, Error>) -> Self {
+        self.post_slot_hook = Some(hook);
+        self
+    }
+
+    /// Compute the state root for `slot` as efficiently as possible.
+    ///
+    /// The `blocks` should be the full list of blocks being applied and `i` should be the index of
+    /// the next block that will be applied, or `blocks.len()` if all blocks have already been
+    /// applied.
+    fn get_state_root(
+        &mut self,
+        slot: Slot,
+        blocks: &[SignedBeaconBlock<E>],
+        i: usize,
+    ) -> Result<Option<Hash256>, Error> {
+        // If we don't care about state roots then return immediately.
+        if self.state_root_strategy == StateRootStrategy::Inconsistent {
+            return Ok(Some(Hash256::zero()));
+        }
+
+        // If a state root iterator is configured, use it to find the root.
+        if let Some(ref mut state_root_iter) = self.state_root_iter {
+            let opt_root = state_root_iter
+                .take_while(|res| res.as_ref().map_or(true, |(_, s)| *s <= slot))
+                .find(|res| res.as_ref().map_or(true, |(_, s)| *s == slot))
+                .transpose()?;
+
+            if let Some((root, _)) = opt_root {
+                return Ok(Some(root));
+            }
+        }
+
+        // Otherwise try to source a root from the previous block.
+        if i > 0 {
+            if let Some(prev_block) = blocks.get(i.safe_sub(1).map_err(BlockReplayError::from)?) {
+                if prev_block.slot() == slot {
+                    return Ok(Some(prev_block.state_root()));
+                }
+            }
+        }
+
+        self.state_root_miss = true;
+        Ok(None)
+    }
+
+    /// Apply `blocks` atop `self.state`, taking care of slot processing.
+    ///
+    /// If `target_slot` is provided then the state will be advanced through to `target_slot`
+    /// after the blocks have been applied.
+    pub fn apply_blocks(
+        mut self,
+        blocks: Vec<SignedBeaconBlock<E>>,
+        target_slot: Option<Slot>,
+    ) -> Result<Self, Error> {
+        for (i, block) in blocks.iter().enumerate() {
+            if block.slot() <= self.state.slot() {
+                continue;
```

Comment:
I would have expected this to raise an error, indicating that the blocks supplied are not valid. It's not a big deal at all, I'm just interested if it's expected to supply out-of-range blocks in some scenario.

Reply:
This was copied over from the old block replay logic in the database, where we load one extra block at the start so that we can read its state root

We could try limiting the number of lead-up blocks to just 1, and error otherwise, but I think it's also acceptable to keep it as-is?

Reply:
I've added a check for `i == 0` so that we'll error if more than 1 lead-up block is supplied